---
title: "Data Metrics"
author: "Yu-Jui Ho and Toby Aicher"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_vignette:
    fig_caption: yes
vignette: >
  %\VignetteIndexEntry{Data Metrics}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


After you successfully upload the data, you will be able to normalize or transform the raw data

## Normalization 

- Reads per millions (RPM) normalization 
- DESeq-like normalization 

## Transformation 

- Variance stablization transformation (VST)
- Log transformation 

After you select the normalization or transformation methods you want to apply to the raw data, you will be able to see how it affects the data. We provide two simple data metrics to help identify samples that deviates from the majority of samples, which may indicate potential problematic samples and can be removed for downstream analysis. 

## Read distribution

It will summerize the total number of read counts for each sample. This is how we control for the bias introduced during sequencing by adjusting the raw count according to their sequencing depth (library size). The read count distribution should be as uniform as possible, and the normalization method you choose will affect the result. 

For RPM normalization, we normalize gene count from each sample according to their total reads (library size). 

For DESeq-like normalization, we use the method introduced from [DESeq](http://bioconductor.org/packages/release/bioc/html/DESeq.html) package. 

A example of how normalization affects the read distribution will look like this 

## Gene coverage

One of the metrics we used to assess the quality of the single-cell library is the `gene drop out rate`. Since the library preparation starts with a small amout of input material, the amplification step is expected to introduced noise based on the reverse transcription and amplification efficiency. Therefore, we summerize the number of gene/transcript that have at least one read covered as a way to evaluate how many genes are being expressed in each sample. This number should be relatively stable across libraries. However, the number of expressed genes may be altered based on the difference between cell types, experimental conditions or sequencing protocals, and intrinsic heterogeneity among cell population, etc. User should be aware of these factors and decide how to treat those outlier samples for downstream analysis.

